### PR TITLE
chore: Add leaderboard definitions

### PR DESCRIPTION
**Pull request by kvcache**
A first step toward a leaderboard offering, I've documented and
translated the internal proposal to a grpc protocol buffers spec.

This isn't connected to any client languages yet; those can be
opted in as we add support and implement the service.

### CODE REVIEW

**Review by virratanasangpunth — DISMISSED** [review]
LGTM

File: proto/leaderboard.proto [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+message ScoreRange {
+  // IEEE 754 single precision 32 bit floating point number.
+  // Momento does not support NaN or Inf in leaderboards.
+  float score_start = 1;
+  // IEEE 754 single precision 32 bit floating point number.
+  // Momento does not support NaN or Inf in leaderboards.
+  float score_end = 2;
+}
+
+message Empty {}
```
Comment (tylerburdsall) [inline proto/leaderboard.proto:134]: @kvcache have you considered using GRPC's default `Empty` message? That way we don't have to have multiple classpaths for an empty message

**Review by tylerburdsall — COMMENTED** [review]
Left a comment

File: proto/leaderboard.proto [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+
+// Query APIs using ScoreRange may match more than the limit of 8192 elements. These apis will
+// include an offset and limit parameter pair, which can be used to page through score ranges
+// matching many elements.
+//
+// ScoreRange models half-open ranges: 0..4 refers to scores 0, 1.1234, 2.5 and 3.999.
+message ScoreRange {
+  // IEEE 754 single precision 32 bit floating point number.
+  // Momento does not support NaN or Inf in leaderboards.
+  float score_start = 1;
```
Comment (krispraws) [inline proto/leaderboard.proto:128]: Nit - naming is inconsistent with RankRange where names also indicate inclusive/exclusive

**Review by tylerburdsall — COMMENTED** [review]
Also I believe this RPCs Request/Response messages should be prefixed with `_`

**Review by krispraws — APPROVED** [review]
Looks good. Left a naming nit comment

File: proto/leaderboard.proto (hunk 1 above)
Comment (kvcache) [reply proto/leaderboard.proto:134]: Yes I did consider that. But this is easier and doesn't have any extra import dependencies. It's also binary-compatible with that message if you ever want to delete this line and replace it with another line that imports this exact structure from someone else's code :-p! This is basically just a more intent-revealing replacement for `XRpcResponse {}` boilerplate that we copypaste and replace X for in many other places.

File: proto/leaderboard.proto (hunk 2 above)
Comment (kvcache) [reply proto/leaderboard.proto:128]: woops thanks

**Comment by kvcache** [comment]
> Also I believe this RPCs Request/Response messages should be prefixed with `_`

Thanks, I forgot this one

**Review by tylerburdsall — APPROVED** [review]

**Review by krispraws — APPROVED** [review]

File: proto/leaderboard.proto [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+  // IEEE 754 single precision 32 bit floating point number.
+  // Momento does not support NaN or Inf in leaderboards.
+  float end_exclusive = 2;
+}
+
+message _Empty {}
+
+message _DeleteLeaderboardRequest {
+  string cache_name = 1;
+  string leaderboard = 2;
```
Comment (virratanasangpunth) [inline proto/leaderboard.proto:138]: nit: should `leaderboard` be `leaderboard_name`? I see we're using `set_name` and `list_name` in the cache proto.

Comment (kvcache) [reply proto/leaderboard.proto:138]: Chatted separately - this ship has kind of sailed at the moment. We can change it later if desired.

**Pull request merged**